### ISSUE TITLE
Parser Error with zero or more parentheses pairs around expression

### ISSUE THREAD

**Issue by TurboThommy** [b3b00/csly#427]
Hi,

while trying to parse a DSL I encountered a problem when using the expression parser.
I need to be able to use parentheses around the expressions but for some reason that does not seem to work.

I tried to reproduce the problem in a small seperate project with following Lexer and Parser:

```
using sly.lexer;

namespace Csly_Playground;

[Lexer]
public enum ExampleTokens
{
    NOT_A_TOKEN,
    [Int] INTEGER,
    [Sugar("(")] LPAREN,
    [Sugar(")")] RPAREN,
    [Sugar("+")] PLUS,
    [Sugar(";")] SEMICOLON,
}
```

```
using sly.lexer;
using sly.parser.generator;

namespace Csly_Playground;

public class ExampleParser
{
    [Production("main : expression")]
    public static int RootTag(int expression) => expression;

    [Operand]
    [Production("operand : INTEGER")]
    public static int IntegerTag(Token<ExampleTokens> expressionToken) => int.Parse(expressionToken.Value.ToString());

    [Operation((int) ExampleTokens.PLUS, Affix.InFix, Associativity.Left, 10)]
    public static int Addition(int left, Token<ExampleTokens> operatorToken, int right) => left + right;

    [Production("expression : LPAREN [d] ExampleParser_expressions RPAREN [d]")]
    [Production("expression : ExampleParser_expressions")]
    public static int Expression (int expression) => expression;
}
```

If I try to parse "2+3" I ecounter an error stating

> unexpected INTEGER ('2 (line 0, column 0)'). Expecting LPAREN, .

Trying to parse "(2+3)" or even "(((2+3)))" works just fine.

I already tried to split the expression rule to two seperate rules with different non-terminals which didn't work as well.
Besides that I tried to change the order of the rule in case the order matters, also without success.

My expectation was, that I would be able to have zero or more parentheses pairs around my expressions.
Is anything wrong with my rules or is there a way around said issue?

**Comment by TurboThommy** [comment]
Alright so when adding another Production rule

```
[Production("expression : LPAREN [d] ExampleParser_expressions RPAREN [d]")]
[Production("expression : ExampleParser_expressions")]
[Production("expression : operand")]
public static int Expression (int expression) => expression;
```

the describes Problem does not occur anymore and inputs like "(2+3+4)" work just fine but now inputs like "(2+3)+4)" don't work. Instead I am getting following error message
> ErrorMessage = "unexpected + (line 0, column 5) PLUS."

Also for inputs like "((2+3)+4)" I get the error message
> unexpected LPAREN ('( (line 0, column 1)'). Expecting INTEGER, .

Is there anything else I need to consider for allowing inputs like "(2+3)+4" as well?

**Comment by b3b00** [comment]
Hello, Thanks for reporting. and thanks to provide a complete reproduction case. i am looking at it right now

**Comment by b3b00** [comment]
@TurboThommy 

Regarding your second message : `(2+3)+4)` is not a valid source , it lacks an opening parenthesis at the begining.
but ((2+3)+4) should parse and evaluate to 10. as (2+3)+4.
Could you clearly restate what works correctly an what does not work, please ? 

With following parser : 

```c# 
public class Issue427Parser
{
    [Production("main : expression")]
    public static int RootTag(int expression) => expression;

    [Operand]
    [Production("operand : INTEGER")]
    public static int IntegerTag(Token<Issue427Lexer> expressionToken) => int.Parse(expressionToken.Value.ToString());

    [Operation((int) Issue427Lexer.PLUS, Affix.InFix, Associativity.Left, 10)]
    public static int Addition(int left, Token<Issue427Lexer> operatorToken, int right) => left + right;

    [Production("expression : LPAREN [d] Issue427Parser_expressions RPAREN [d]")]
    [Production("expression : Issue427Parser_expressions")]
    [Production("expression : operand")]
    public static int Expression (int expression) => expression;
}
```

I have : 

-  `2 + 3` : KO : unexpected end of stream. Expecting  (line 0, column 5).
- `(2+3)` : KO : unexpected RPAREN (') (line 0, column 6)'). Expecting PLUS, .
- `((2+3)+4)`  :  unexpected LPAREN ('( (line 0, column 1)'). Expecting INTEGER, .
- `(2+3)+4)` : unexpected RPAREN (') (line 0, column 4)'). Expecting PLUS, . This is expected as the RPAREN has no matching LPAREN

**Comment by b3b00** [comment]
At least for expression `2 + 3` I've reduced the reproduction grammar to this : 

```c#
public class Issue427MinimalParser
{

    [Production("root : Issue427MinimalParser_expressions")]
    public static int Root(int value) => value;

    [Operand]
    [Production("operand : INTEGER")]
    public static int operand(Token<Issue427Lexer> expressionToken) => expressionToken.IntValue;

    [Operation((int) Issue427Lexer.PLUS, Affix.InFix, Associativity.Left, 10)]
    public static int Addition(int left, Token<Issue427Lexer> operatorToken, int right) => left + right;

}
```

The issue is related to the expression parser generator feature. Still investigating.

**Comment by b3b00** [comment]
@TurboThommy , I've published a new nuget [2.9.9](https://www.nuget.org/packages/sly) with a fix. You must nevertheless correct your parser as follow :

```c#
public class Issue427BisParser
{
    [Production("main : expression")]
    public static int RootTag(int expression) => expression;

    [Operand]
    [Production("operand : INTEGER")]
    public static int IntegerTag(Token<Issue427Lexer> expressionToken) => int.Parse(expressionToken.Value.ToString());

    // a parenthesised expression can be an operand for another expression
    [Production("group : LPAREN [d] Issue427BisParser_expressions RPAREN [d]")]
    [Operand]
    public static int Group(int value) => value;

    [Operation((int) Issue427Lexer.PLUS, Affix.InFix, Associativity.Left, 10)]
    public static int Addition(int left, Token<Issue427Lexer> operatorToken, int right) => left + right;
    
    [Production("expression : Issue427BisParser_expressions")]
    public static int Expression (int expression) => expression;
}
```

you can see full example and unit tests here : https://github.com/b3b00/csly/blob/dev/tests/ParserTests/issue427/Issue427Tests.cs

let me know if it is ok for you

**Comment by TurboThommy** [comment]
Hello @b3b00, thanks a lot. 

You are absolutely right, there was a typo in the example I have given.
After updating the nuget package and adjusting my production rules according to your last message everything works like a charm.
Thanks a lot!
